parlai/agents/bart/bart.py

```diff
@@ -0,0 +1,181 @@
+from parlai.agents.transformer.transformer import TransformerGeneratorAgent
+from parlai.core.agents import compare_init_model_opts
+from parlai.core.message import Message
+from parlai.core.opt import Opt
+from parlai.core.params import ParlaiParser
+from parlai.core.torch_agent import Batch, History
+from parlai.utils.typing import TShared
+from parlai.zoo.bart.build import download, CONVERSION_ARGS, BART_ARGS
+
+
+class BartAgent(TransformerGeneratorAgent):
+    """
+    BART Agent.
+
+    Relies on the BART model implemented in fairseq.
+
+    If you have a fine-tuned BART model from fairseq, you can specify the
+    `--init-fairseq-model` arg, which will convert your fine-tuned model
+    to a ParlAI model.
+    """
+
+    @staticmethod
+    def add_cmdline_args(argparser: ParlaiParser):
+        """
+        Override to add init-fairseq-model arg.
+        """
+        TransformerGeneratorAgent.add_cmdline_args(argparser)
+        group = argparser.add_argument_group('Bart Args')
+        group.add_argument(
+            '--init-fairseq-model',
+            type=str,
+            default=None,
+            help='fairseq checkpoint for bart',
+        )
+        argparser.set_defaults(dict_tokenizer='gpt2')
+
+    def __init__(self, opt: Opt, shared: TShared = None):
+        if not shared:
+            opt = self._initialize_bart(opt)
+        super().__init__(opt, shared)
+
+    def _initialize_bart(self, opt: Opt) -> Opt:
+        """
+        Download and convert BART pre-trained models.
+
+        Additionally, convert `init-fairseq-model` if necessary.
+
+        :param opt:
+            ParlAI-parsed options
+
+        :return opt:
+            return opt with BART-specific args.
+        """
+        if not opt.get('converting'):
+            download(opt['datapath'])
+            opt['init_model'] = os.path.join(
+                opt['datapath'], 'models/bart/bart_large/model'
+            )
+        if opt.get('init_fairseq_model'):
+            opt = self._convert_model(opt)
+        opt.update(BART_ARGS)
+        compare_init_model_opts(opt, opt)
+        return opt
+
+    def _convert_model(self, opt: Opt) -> Opt:
+        """
+        Convert fairseq init model to ParlAI Model.
+
+        :param opt:
+            options
+
+        :return opt:
+            return opt with new init_model path
+        """
+        model_name = os.path.split(opt['init_fairseq_model'])[-1]
+        args = CONVERSION_ARGS.copy()
+
+        args['input'] = [opt['init_fairseq_model']]
+        if opt.get('model_file') and not os.path.exists(opt['model_file']):
+            args['output'] = opt['model_file']
+        else:
+            args['output'] = os.path.join(
+                opt['datapath'], 'models/converted_fairseq_models/', model_name
+            )
+
+        ConversionScript.main(**args)
+        opt['init_model'] = args['output']
+        return opt
+
+    def build_model(self, states: Any = None) -> BartModel:
```

Comment:
does build_model ever take states?

Reply:
nope, actually mypy complained this didn't have the same signature as it's superclass (`TransformerGeneratorAgent`) which is why I added it back. 

probably better to just remove from both